certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -250,6 +259,121 @@ def _get_ranked_matches(self, target_name):
                                 'rank': 6 if vhost.ssl else 7})
         return sorted(matches, key=lambda x: x['rank'])
 
+    def all_vhost_ports(self, domain):
+        """Get all ports that all vhosts are listening on, that match domain.
+
+        :param str domain: domain to find ports of
+        :returns: set of strings representing the port numbers
+        :rtype: set
+        """
+        matching_vhosts = [item['vhost'] for item in self._get_ranked_matches(domain)]
+        out = set([addr.get_port() for vhost in matching_vhosts for addr in vhost.addrs])
```

Comment:
Making this change doesn't break anything. It's certainly not a necessary change, but it should be a little faster.

You probably have used generators without even knowing it! `range` returns a list. `xrange` returns a generator. No real gotchas. Closest thing to a gotcha is you can only iterate forward through generator expressions and you can only go through them once. None of that matters here.